{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from joblib import Parallel, delayed\n",
    "from utils import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# 1. Loop\n",
    "\n",
    "For-loop through VCF records and use functions from `utils.py` to extract annotation information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record(CHROM=1, POS=1158631, REF=A, ALT=[G])\n",
      "\n",
      "Record(CHROM=1, POS=1246004, REF=A, ALT=[G])\n",
      "\n",
      "Record(CHROM=1, POS=1249187, REF=G, ALT=[A])\n",
      "\n",
      "Record(CHROM=1, POS=1261824, REF=G, ALT=[C])\n",
      "\n",
      "Record(CHROM=1, POS=1387667, REF=C, ALT=[G])\n",
      "\n",
      "Record(CHROM=1, POS=1585597, REF=A, ALT=[G])\n",
      "\n",
      "Record(CHROM=1, POS=1585642, REF=G, ALT=[T])\n",
      "\n",
      "Record(CHROM=1, POS=1586752, REF=T, ALT=[C])\n",
      "\n",
      "Record(CHROM=1, POS=1647686, REF=A, ALT=[C])\n",
      "\n",
      "Record(CHROM=1, POS=1647722, REF=GCTGTGACA, ALT=[TCTAGGATG])\n",
      "\n",
      "Record(CHROM=1, POS=1647745, REF=GGCCCTTTC, ALT=[AGCCCTTTT])\n",
      "\n",
      "Record(CHROM=1, POS=1647778, REF=C, ALT=[G])\n",
      "\n",
      "Record(CHROM=1, POS=1647893, REF=C, ALT=[CTTTCTT])\n",
      "\n",
      "Record(CHROM=1, POS=1647968, REF=C, ALT=[CAT])\n",
      "\n",
      "Record(CHROM=1, POS=1647971, REF=G, ALT=[A])\n",
      "\n",
      "Record(CHROM=1, POS=1647983, REF=TGGCTTAC, ALT=[AGGCTTAT])\n",
      "\n",
      "Record(CHROM=1, POS=1650787, REF=TGATGCCTACATTTT, ALT=[CGATGCCTACGTTTC])\n",
      "\n",
      "Record(CHROM=1, POS=1650807, REF=T, ALT=[C])\n",
      "\n",
      "Record(CHROM=1, POS=1650832, REF=A, ALT=[G])\n",
      "\n",
      "Record(CHROM=1, POS=1650845, REF=G, ALT=[A])\n",
      "\n",
      "Record(CHROM=1, POS=1885055, REF=C, ALT=[T])\n",
      "\n",
      "Record(CHROM=1, POS=1895177, REF=C, ALT=[T])\n",
      "\n",
      "Record(CHROM=1, POS=1900106, REF=T, ALT=[TCTC])\n",
      "\n",
      "Record(CHROM=1, POS=1900232, REF=T, ALT=[C])\n",
      "\n",
      "Record(CHROM=1, POS=1909843, REF=C, ALT=[T])\n",
      "\n",
      "Record(CHROM=1, POS=1909854, REF=G, ALT=[A])\n",
      "\n",
      "Record(CHROM=1, POS=1909868, REF=G, ALT=[A])\n",
      "\n",
      "Record(CHROM=1, POS=1910001, REF=C, ALT=[G])\n",
      "\n",
      "Record(CHROM=1, POS=1910012, REF=T, ALT=[C])\n",
      "\n",
      "Record(CHROM=1, POS=2328714, REF=A, ALT=[G])\n",
      "\n",
      "Record(CHROM=1, POS=3638674, REF=C, ALT=[T])\n",
      "\n",
      "Record(CHROM=1, POS=3679775, REF=C, ALT=[T])\n",
      "\n",
      "Record(CHROM=1, POS=3743109, REF=A, ALT=[G])\n",
      "\n",
      "Record(CHROM=1, POS=3743132, REF=T, ALT=[C])\n",
      "\n",
      "Record(CHROM=1, POS=3743391, REF=C, ALT=[T])\n",
      "\n",
      "Record(CHROM=1, POS=3745787, REF=T, ALT=[C])\n",
      "\n",
      "Record(CHROM=1, POS=3755675, REF=T, ALT=[C])\n",
      "\n",
      "Record(CHROM=1, POS=3764216, REF=C, ALT=[T])\n",
      "\n",
      "Record(CHROM=1, POS=6158562, REF=A, ALT=[G])\n",
      "\n",
      "Record(CHROM=1, POS=6219287, REF=TCACACA, ALT=[TCACA, T])\n",
      "\n",
      "\n",
      "Done! Total records: 40\n",
      "CPU times: user 1.3 s, sys: 40.5 ms, total: 1.34 s\n",
      "Wall time: 1min 17s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "out_df = pd.DataFrame()  # initialize empty dataframe\n",
    "vcf_read = vcf.Reader(open(\"test_vcf_data.txt\", \"r\"))\n",
    "\n",
    "for i, record in enumerate(vcf_read):\n",
    "    if i < 40:  # only look through first few records for manageable output\n",
    "        print(record, end = \"\\n\\n\")\n",
    "\n",
    "        # unpack record into dictionary of depth 1\n",
    "        record_unpacked = unpack_dict(record.__dict__)\n",
    "        \n",
    "        # unpack ALT allele objects\n",
    "        record_unpacked = extract_alt(record_unpacked, alt_obj=record.ALT, alt_key=\"ALT\")\n",
    "            \n",
    "        # unpack model._Call object\n",
    "        call_dict = unpack_dict(\n",
    "            extract_call(record_unpacked[\"samples\"], format_str=record_unpacked[\"FORMAT\"]),\n",
    "        )\n",
    "        # remove old keys\n",
    "        del record_unpacked[\"samples\"]\n",
    "        del record_unpacked[\"FORMAT\"]\n",
    "        # merge dicts\n",
    "        record_unpacked = {**record_unpacked, **call_dict}\n",
    "    \n",
    "        # get gene where mutation is\n",
    "        overlap_dict = overlap_get(record)\n",
    "        try:\n",
    "            # first record\n",
    "            record_unpacked[\"gene_id\"] = overlap_dict[0][\"gene_id\"]\n",
    "            record_unpacked[\"gene_symbol\"] = overlap_dict[0][\"external_name\"]\n",
    "            record_unpacked[\"gene_biotype\"] = overlap_dict[0][\"biotype\"]\n",
    "            record_unpacked[\"gene_description\"] = overlap_dict[0][\"description\"]\n",
    "            if len(overlap_dict) > 1:\n",
    "                for i in range(1, len(overlap_dict)):\n",
    "                    record_unpacked[\"gene_id{}\".format(i)] = overlap_dict[i][\"gene_id\"]\n",
    "                    record_unpacked[\"gene_symbol{}\".format(i)] = overlap_dict[i][\"external_name\"]\n",
    "                    record_unpacked[\"gene_biotype{}\".format(i)] = overlap_dict[i][\"biotype\"]\n",
    "                    record_unpacked[\"gene_description{}\".format(i)] = overlap_dict[i][\"description\"]\n",
    "        except:\n",
    "            print(\"No gene overlap found for {}\".format(record))\n",
    "\n",
    "        # call Ensembl VEP endpoint\n",
    "        vep_dict = vep_region_post(\n",
    "            record,\n",
    "            hgvs=1,\n",
    "            #protein=1,\n",
    "            #uniprot=1,\n",
    "            #GO=1,\n",
    "            #vcf_string=1,\n",
    "            #domains=1,\n",
    "            distance=0,  # don't look for upstream/downstream effects for simplicity\n",
    "        )\n",
    "        # extract VEP info\n",
    "        record_unpacked[\"most_severe_consequence\"] = vep_dict[\"most_severe_consequence\"]\n",
    "        \n",
    "        # get MAF and snp info if available\n",
    "        if \"colocated_variants\" in vep_dict:\n",
    "            for d in vep_dict[\"colocated_variants\"]:\n",
    "                if \"minor_allele\" in d:\n",
    "                    record_unpacked[\"minor_allele\"] = d[\"minor_allele\"]\n",
    "                    record_unpacked[\"minor_allele_freq\"] = d[\"minor_allele_freq\"]\n",
    "                    record_unpacked[\"snp_id\"] = d[\"id\"]\n",
    "        \n",
    "        # add to out_df\n",
    "        out_df = pd.concat([out_df, pd.DataFrame(record_unpacked, index=[i])])\n",
    "    else:\n",
    "        break  # after n loops, quit to look at output\n",
    "        \n",
    "calc_frequencies(out_df)  # calculate allele frequencies\n",
    "    \n",
    "print(\"\\nDone! Total records: {}\".format(i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ALT_type</th>\n",
       "      <th>NR</th>\n",
       "      <th>NV</th>\n",
       "      <th>NRef</th>\n",
       "      <th>VAF</th>\n",
       "      <th>RAF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1158631</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>160</td>\n",
       "      <td>156</td>\n",
       "      <td>4</td>\n",
       "      <td>0.975</td>\n",
       "      <td>0.025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1246004</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>152</td>\n",
       "      <td>148</td>\n",
       "      <td>4</td>\n",
       "      <td>0.973684</td>\n",
       "      <td>0.026316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1249187</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>137</td>\n",
       "      <td>135</td>\n",
       "      <td>2</td>\n",
       "      <td>0.985401</td>\n",
       "      <td>0.014599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1261824</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>136</td>\n",
       "      <td>134</td>\n",
       "      <td>2</td>\n",
       "      <td>0.985294</td>\n",
       "      <td>0.014706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1387667</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>137</td>\n",
       "      <td>133</td>\n",
       "      <td>4</td>\n",
       "      <td>0.970803</td>\n",
       "      <td>0.029197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>1585597</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>264</td>\n",
       "      <td>123</td>\n",
       "      <td>141</td>\n",
       "      <td>0.465909</td>\n",
       "      <td>0.534091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>1585642</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>196</td>\n",
       "      <td>96</td>\n",
       "      <td>100</td>\n",
       "      <td>0.489796</td>\n",
       "      <td>0.510204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>1586752</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>189</td>\n",
       "      <td>186</td>\n",
       "      <td>3</td>\n",
       "      <td>0.984127</td>\n",
       "      <td>0.015873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1647686</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>178</td>\n",
       "      <td>131</td>\n",
       "      <td>47</td>\n",
       "      <td>0.735955</td>\n",
       "      <td>0.264045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1647722</td>\n",
       "      <td>GCTGTGACA</td>\n",
       "      <td>TCTAGGATG</td>\n",
       "      <td>MNV</td>\n",
       "      <td>259</td>\n",
       "      <td>60</td>\n",
       "      <td>199</td>\n",
       "      <td>0.23166</td>\n",
       "      <td>0.76834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1647745</td>\n",
       "      <td>GGCCCTTTC</td>\n",
       "      <td>AGCCCTTTT</td>\n",
       "      <td>MNV</td>\n",
       "      <td>305</td>\n",
       "      <td>74</td>\n",
       "      <td>231</td>\n",
       "      <td>0.242623</td>\n",
       "      <td>0.757377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1647778</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>485</td>\n",
       "      <td>154</td>\n",
       "      <td>331</td>\n",
       "      <td>0.317526</td>\n",
       "      <td>0.682474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1647893</td>\n",
       "      <td>C</td>\n",
       "      <td>CTTTCTT</td>\n",
       "      <td>MNV</td>\n",
       "      <td>379</td>\n",
       "      <td>60</td>\n",
       "      <td>319</td>\n",
       "      <td>0.158311</td>\n",
       "      <td>0.841689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1647968</td>\n",
       "      <td>C</td>\n",
       "      <td>CAT</td>\n",
       "      <td>MNV</td>\n",
       "      <td>175</td>\n",
       "      <td>51</td>\n",
       "      <td>124</td>\n",
       "      <td>0.291429</td>\n",
       "      <td>0.708571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1647971</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>162</td>\n",
       "      <td>47</td>\n",
       "      <td>115</td>\n",
       "      <td>0.290123</td>\n",
       "      <td>0.709877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1647983</td>\n",
       "      <td>TGGCTTAC</td>\n",
       "      <td>AGGCTTAT</td>\n",
       "      <td>MNV</td>\n",
       "      <td>131</td>\n",
       "      <td>37</td>\n",
       "      <td>94</td>\n",
       "      <td>0.282443</td>\n",
       "      <td>0.717557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1650787</td>\n",
       "      <td>TGATGCCTACATTTT</td>\n",
       "      <td>CGATGCCTACGTTTC</td>\n",
       "      <td>MNV</td>\n",
       "      <td>281</td>\n",
       "      <td>99</td>\n",
       "      <td>182</td>\n",
       "      <td>0.352313</td>\n",
       "      <td>0.647687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1650807</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>303</td>\n",
       "      <td>165</td>\n",
       "      <td>138</td>\n",
       "      <td>0.544554</td>\n",
       "      <td>0.455446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1650832</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>314</td>\n",
       "      <td>151</td>\n",
       "      <td>163</td>\n",
       "      <td>0.480892</td>\n",
       "      <td>0.519108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1650845</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>320</td>\n",
       "      <td>160</td>\n",
       "      <td>160</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>1885055</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>105</td>\n",
       "      <td>105</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1</td>\n",
       "      <td>1895177</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>108</td>\n",
       "      <td>107</td>\n",
       "      <td>1</td>\n",
       "      <td>0.990741</td>\n",
       "      <td>0.009259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1</td>\n",
       "      <td>1900106</td>\n",
       "      <td>T</td>\n",
       "      <td>TCTC</td>\n",
       "      <td>MNV</td>\n",
       "      <td>147</td>\n",
       "      <td>123</td>\n",
       "      <td>24</td>\n",
       "      <td>0.836735</td>\n",
       "      <td>0.163265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>1900232</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>247</td>\n",
       "      <td>245</td>\n",
       "      <td>2</td>\n",
       "      <td>0.991903</td>\n",
       "      <td>0.008097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1</td>\n",
       "      <td>1909843</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>212</td>\n",
       "      <td>200</td>\n",
       "      <td>12</td>\n",
       "      <td>0.943396</td>\n",
       "      <td>0.056604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1</td>\n",
       "      <td>1909854</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>220</td>\n",
       "      <td>210</td>\n",
       "      <td>10</td>\n",
       "      <td>0.954545</td>\n",
       "      <td>0.045455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1</td>\n",
       "      <td>1909868</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>226</td>\n",
       "      <td>218</td>\n",
       "      <td>8</td>\n",
       "      <td>0.964602</td>\n",
       "      <td>0.035398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1</td>\n",
       "      <td>1910001</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>142</td>\n",
       "      <td>140</td>\n",
       "      <td>2</td>\n",
       "      <td>0.985915</td>\n",
       "      <td>0.014085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1</td>\n",
       "      <td>1910012</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>128</td>\n",
       "      <td>127</td>\n",
       "      <td>1</td>\n",
       "      <td>0.992188</td>\n",
       "      <td>0.007812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1</td>\n",
       "      <td>2328714</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>136</td>\n",
       "      <td>133</td>\n",
       "      <td>3</td>\n",
       "      <td>0.977941</td>\n",
       "      <td>0.022059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1</td>\n",
       "      <td>3638674</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>105</td>\n",
       "      <td>98</td>\n",
       "      <td>7</td>\n",
       "      <td>0.933333</td>\n",
       "      <td>0.066667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1</td>\n",
       "      <td>3679775</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>123</td>\n",
       "      <td>115</td>\n",
       "      <td>8</td>\n",
       "      <td>0.934959</td>\n",
       "      <td>0.065041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1</td>\n",
       "      <td>3743109</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>155</td>\n",
       "      <td>151</td>\n",
       "      <td>4</td>\n",
       "      <td>0.974194</td>\n",
       "      <td>0.025806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1</td>\n",
       "      <td>3743132</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>180</td>\n",
       "      <td>178</td>\n",
       "      <td>2</td>\n",
       "      <td>0.988889</td>\n",
       "      <td>0.011111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1</td>\n",
       "      <td>3743391</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>153</td>\n",
       "      <td>152</td>\n",
       "      <td>1</td>\n",
       "      <td>0.993464</td>\n",
       "      <td>0.006536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1</td>\n",
       "      <td>3745787</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>105</td>\n",
       "      <td>105</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1</td>\n",
       "      <td>3755675</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>101</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>0.990099</td>\n",
       "      <td>0.009901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1</td>\n",
       "      <td>3764216</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>101</td>\n",
       "      <td>101</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>1</td>\n",
       "      <td>6158562</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>262</td>\n",
       "      <td>125</td>\n",
       "      <td>137</td>\n",
       "      <td>0.477099</td>\n",
       "      <td>0.522901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>1</td>\n",
       "      <td>6219287</td>\n",
       "      <td>TCACACA</td>\n",
       "      <td>TCACA;T</td>\n",
       "      <td>MNV;SNV</td>\n",
       "      <td>172;172</td>\n",
       "      <td>108;108</td>\n",
       "      <td>64;64.0</td>\n",
       "      <td>0.627906976744186;0.627906976744186</td>\n",
       "      <td>0.37209302325581395;0.37209302325581395</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM      POS              REF              ALT ALT_type       NR  \\\n",
       "0      1  1158631                A                G      SNV      160   \n",
       "1      1  1246004                A                G      SNV      152   \n",
       "2      1  1249187                G                A      SNV      137   \n",
       "3      1  1261824                G                C      SNV      136   \n",
       "4      1  1387667                C                G      SNV      137   \n",
       "5      1  1585597                A                G      SNV      264   \n",
       "6      1  1585642                G                T      SNV      196   \n",
       "7      1  1586752                T                C      SNV      189   \n",
       "1      1  1647686                A                C      SNV      178   \n",
       "1      1  1647722        GCTGTGACA        TCTAGGATG      MNV      259   \n",
       "1      1  1647745        GGCCCTTTC        AGCCCTTTT      MNV      305   \n",
       "1      1  1647778                C                G      SNV      485   \n",
       "1      1  1647893                C          CTTTCTT      MNV      379   \n",
       "1      1  1647968                C              CAT      MNV      175   \n",
       "1      1  1647971                G                A      SNV      162   \n",
       "1      1  1647983         TGGCTTAC         AGGCTTAT      MNV      131   \n",
       "1      1  1650787  TGATGCCTACATTTT  CGATGCCTACGTTTC      MNV      281   \n",
       "1      1  1650807                T                C      SNV      303   \n",
       "1      1  1650832                A                G      SNV      314   \n",
       "1      1  1650845                G                A      SNV      320   \n",
       "20     1  1885055                C                T      SNV      105   \n",
       "21     1  1895177                C                T      SNV      108   \n",
       "22     1  1900106                T             TCTC      MNV      147   \n",
       "23     1  1900232                T                C      SNV      247   \n",
       "24     1  1909843                C                T      SNV      212   \n",
       "25     1  1909854                G                A      SNV      220   \n",
       "26     1  1909868                G                A      SNV      226   \n",
       "27     1  1910001                C                G      SNV      142   \n",
       "28     1  1910012                T                C      SNV      128   \n",
       "29     1  2328714                A                G      SNV      136   \n",
       "30     1  3638674                C                T      SNV      105   \n",
       "31     1  3679775                C                T      SNV      123   \n",
       "32     1  3743109                A                G      SNV      155   \n",
       "33     1  3743132                T                C      SNV      180   \n",
       "34     1  3743391                C                T      SNV      153   \n",
       "35     1  3745787                T                C      SNV      105   \n",
       "36     1  3755675                T                C      SNV      101   \n",
       "37     1  3764216                C                T      SNV      101   \n",
       "38     1  6158562                A                G      SNV      262   \n",
       "39     1  6219287          TCACACA          TCACA;T  MNV;SNV  172;172   \n",
       "\n",
       "         NV     NRef                                  VAF  \\\n",
       "0       156        4                                0.975   \n",
       "1       148        4                             0.973684   \n",
       "2       135        2                             0.985401   \n",
       "3       134        2                             0.985294   \n",
       "4       133        4                             0.970803   \n",
       "5       123      141                             0.465909   \n",
       "6        96      100                             0.489796   \n",
       "7       186        3                             0.984127   \n",
       "1       131       47                             0.735955   \n",
       "1        60      199                              0.23166   \n",
       "1        74      231                             0.242623   \n",
       "1       154      331                             0.317526   \n",
       "1        60      319                             0.158311   \n",
       "1        51      124                             0.291429   \n",
       "1        47      115                             0.290123   \n",
       "1        37       94                             0.282443   \n",
       "1        99      182                             0.352313   \n",
       "1       165      138                             0.544554   \n",
       "1       151      163                             0.480892   \n",
       "1       160      160                                  0.5   \n",
       "20      105        0                                  1.0   \n",
       "21      107        1                             0.990741   \n",
       "22      123       24                             0.836735   \n",
       "23      245        2                             0.991903   \n",
       "24      200       12                             0.943396   \n",
       "25      210       10                             0.954545   \n",
       "26      218        8                             0.964602   \n",
       "27      140        2                             0.985915   \n",
       "28      127        1                             0.992188   \n",
       "29      133        3                             0.977941   \n",
       "30       98        7                             0.933333   \n",
       "31      115        8                             0.934959   \n",
       "32      151        4                             0.974194   \n",
       "33      178        2                             0.988889   \n",
       "34      152        1                             0.993464   \n",
       "35      105        0                                  1.0   \n",
       "36      100        1                             0.990099   \n",
       "37      101        0                                  1.0   \n",
       "38      125      137                             0.477099   \n",
       "39  108;108  64;64.0  0.627906976744186;0.627906976744186   \n",
       "\n",
       "                                        RAF  \n",
       "0                                     0.025  \n",
       "1                                  0.026316  \n",
       "2                                  0.014599  \n",
       "3                                  0.014706  \n",
       "4                                  0.029197  \n",
       "5                                  0.534091  \n",
       "6                                  0.510204  \n",
       "7                                  0.015873  \n",
       "1                                  0.264045  \n",
       "1                                   0.76834  \n",
       "1                                  0.757377  \n",
       "1                                  0.682474  \n",
       "1                                  0.841689  \n",
       "1                                  0.708571  \n",
       "1                                  0.709877  \n",
       "1                                  0.717557  \n",
       "1                                  0.647687  \n",
       "1                                  0.455446  \n",
       "1                                  0.519108  \n",
       "1                                       0.5  \n",
       "20                                      0.0  \n",
       "21                                 0.009259  \n",
       "22                                 0.163265  \n",
       "23                                 0.008097  \n",
       "24                                 0.056604  \n",
       "25                                 0.045455  \n",
       "26                                 0.035398  \n",
       "27                                 0.014085  \n",
       "28                                 0.007812  \n",
       "29                                 0.022059  \n",
       "30                                 0.066667  \n",
       "31                                 0.065041  \n",
       "32                                 0.025806  \n",
       "33                                 0.011111  \n",
       "34                                 0.006536  \n",
       "35                                      0.0  \n",
       "36                                 0.009901  \n",
       "37                                      0.0  \n",
       "38                                 0.522901  \n",
       "39  0.37209302325581395;0.37209302325581395  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_df[[\"CHROM\",\"POS\",\"REF\",\"ALT\",\"ALT_type\",\"NR\",\"NV\",\"NRef\",\"VAF\",\"RAF\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'intron_variant'"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vep_dict[\"most_severe_consequence\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CHD5', 'CHD5', 'CHD5']"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[vep_dict[\"transcript_consequences\"][x][\"gene_symbol\"] for x in range(len(vep_dict[\"transcript_consequences\"]))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['protein_coding', 'protein_coding', 'nonsense_mediated_decay']"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[vep_dict[\"transcript_consequences\"][x][\"biotype\"] for x in range(len(vep_dict[\"transcript_consequences\"]))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['intron_variant'],\n",
       " ['intron_variant'],\n",
       " ['intron_variant', 'NMD_transcript_variant']]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[vep_dict[\"transcript_consequences\"][x][\"consequence_terms\"] for x in range(len(vep_dict[\"transcript_consequences\"]))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'gene_id': 'ENSG00000116254',\n",
       "  'variant_allele': 'CACA',\n",
       "  'transcript_id': 'ENST00000262450',\n",
       "  'biotype': 'protein_coding',\n",
       "  'hgvsc': 'ENST00000262450.3:c.387+107_387+108del',\n",
       "  'hgnc_id': 16816,\n",
       "  'gene_symbol': 'CHD5',\n",
       "  'strand': -1,\n",
       "  'consequence_terms': ['intron_variant'],\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'impact': 'MODIFIER'},\n",
       " {'transcript_id': 'ENST00000378021',\n",
       "  'variant_allele': 'CACA',\n",
       "  'gene_id': 'ENSG00000116254',\n",
       "  'biotype': 'protein_coding',\n",
       "  'hgvsc': 'ENST00000378021.1:c.-3043+107_-3043+108del',\n",
       "  'hgnc_id': 16816,\n",
       "  'strand': -1,\n",
       "  'consequence_terms': ['intron_variant'],\n",
       "  'gene_symbol': 'CHD5',\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'impact': 'MODIFIER'},\n",
       " {'variant_allele': 'CACA',\n",
       "  'transcript_id': 'ENST00000496404',\n",
       "  'gene_id': 'ENSG00000116254',\n",
       "  'biotype': 'nonsense_mediated_decay',\n",
       "  'hgnc_id': 16816,\n",
       "  'hgvsc': 'ENST00000496404.1:c.387+107_387+108del',\n",
       "  'impact': 'MODIFIER',\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'strand': -1,\n",
       "  'consequence_terms': ['intron_variant', 'NMD_transcript_variant'],\n",
       "  'gene_symbol': 'CHD5'}]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vep_dict[\"transcript_consequences\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# 2. `joblib` Parallel\n",
    "\n",
    "Try running in parallel to speed it up."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 29.2 s, sys: 1.47 s, total: 30.7 s\n",
      "Wall time: 1h 4min 47s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "vcf_read = vcf.Reader(open(\"test_vcf_data.txt\", \"r\"))\n",
    "out = Parallel(n_jobs=5)(delayed(process_record)(record, i) for i, record in enumerate(vcf_read))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11765"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(out)  # list of pd.DataFrames returned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_df = pd.concat(out)  # concatenate into master df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>alleles</th>\n",
       "      <th>affected_start</th>\n",
       "      <th>affected_end</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_biotype5</th>\n",
       "      <th>gene_description5</th>\n",
       "      <th>gene_id6</th>\n",
       "      <th>gene_symbol6</th>\n",
       "      <th>gene_biotype6</th>\n",
       "      <th>gene_description6</th>\n",
       "      <th>gene_id7</th>\n",
       "      <th>gene_symbol7</th>\n",
       "      <th>gene_biotype7</th>\n",
       "      <th>gene_description7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1158631</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2965</td>\n",
       "      <td>1158630</td>\n",
       "      <td>1158631</td>\n",
       "      <td>A;G</td>\n",
       "      <td>1158630</td>\n",
       "      <td>1158631</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1246004</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2965</td>\n",
       "      <td>1246003</td>\n",
       "      <td>1246004</td>\n",
       "      <td>A;G</td>\n",
       "      <td>1246003</td>\n",
       "      <td>1246004</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1249187</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2965</td>\n",
       "      <td>1249186</td>\n",
       "      <td>1249187</td>\n",
       "      <td>G;A</td>\n",
       "      <td>1249186</td>\n",
       "      <td>1249187</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1261824</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>2965</td>\n",
       "      <td>1261823</td>\n",
       "      <td>1261824</td>\n",
       "      <td>G;C</td>\n",
       "      <td>1261823</td>\n",
       "      <td>1261824</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1387667</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2965</td>\n",
       "      <td>1387666</td>\n",
       "      <td>1387667</td>\n",
       "      <td>C;G</td>\n",
       "      <td>1387666</td>\n",
       "      <td>1387667</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11760</th>\n",
       "      <td>X</td>\n",
       "      <td>154020114</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>2965</td>\n",
       "      <td>154020113</td>\n",
       "      <td>154020114</td>\n",
       "      <td>C;A</td>\n",
       "      <td>154020113</td>\n",
       "      <td>154020114</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11761</th>\n",
       "      <td>X</td>\n",
       "      <td>154456747</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2965</td>\n",
       "      <td>154456746</td>\n",
       "      <td>154456747</td>\n",
       "      <td>A;G</td>\n",
       "      <td>154456746</td>\n",
       "      <td>154456747</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11762</th>\n",
       "      <td>X</td>\n",
       "      <td>155125435</td>\n",
       "      <td>AG</td>\n",
       "      <td>A</td>\n",
       "      <td>2973</td>\n",
       "      <td>155125434</td>\n",
       "      <td>155125436</td>\n",
       "      <td>AG;A</td>\n",
       "      <td>155125435</td>\n",
       "      <td>155125436</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11763</th>\n",
       "      <td>X</td>\n",
       "      <td>155127675</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2965</td>\n",
       "      <td>155127674</td>\n",
       "      <td>155127675</td>\n",
       "      <td>A;G</td>\n",
       "      <td>155127674</td>\n",
       "      <td>155127675</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11764</th>\n",
       "      <td>X</td>\n",
       "      <td>155233098</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>2962</td>\n",
       "      <td>155233097</td>\n",
       "      <td>155233098</td>\n",
       "      <td>T;C</td>\n",
       "      <td>155233097</td>\n",
       "      <td>155233098</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11765 rows Ã— 71 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM        POS REF ALT  QUAL      start        end alleles  \\\n",
       "0         1    1158631   A   G  2965    1158630    1158631     A;G   \n",
       "1         1    1246004   A   G  2965    1246003    1246004     A;G   \n",
       "2         1    1249187   G   A  2965    1249186    1249187     G;A   \n",
       "3         1    1261824   G   C  2965    1261823    1261824     G;C   \n",
       "4         1    1387667   C   G  2965    1387666    1387667     C;G   \n",
       "...     ...        ...  ..  ..   ...        ...        ...     ...   \n",
       "11760     X  154020114   C   A  2965  154020113  154020114     C;A   \n",
       "11761     X  154456747   A   G  2965  154456746  154456747     A;G   \n",
       "11762     X  155125435  AG   A  2973  155125434  155125436    AG;A   \n",
       "11763     X  155127675   A   G  2965  155127674  155127675     A;G   \n",
       "11764     X  155233098   T   C  2962  155233097  155233098     T;C   \n",
       "\n",
       "       affected_start  affected_end  ...  gene_biotype5 gene_description5  \\\n",
       "0             1158630       1158631  ...            NaN               NaN   \n",
       "1             1246003       1246004  ...            NaN               NaN   \n",
       "2             1249186       1249187  ...            NaN               NaN   \n",
       "3             1261823       1261824  ...            NaN               NaN   \n",
       "4             1387666       1387667  ...            NaN               NaN   \n",
       "...               ...           ...  ...            ...               ...   \n",
       "11760       154020113     154020114  ...            NaN               NaN   \n",
       "11761       154456746     154456747  ...            NaN               NaN   \n",
       "11762       155125435     155125436  ...            NaN               NaN   \n",
       "11763       155127674     155127675  ...            NaN               NaN   \n",
       "11764       155233097     155233098  ...            NaN               NaN   \n",
       "\n",
       "       gene_id6  gene_symbol6  gene_biotype6  gene_description6  gene_id7  \\\n",
       "0           NaN           NaN            NaN                NaN       NaN   \n",
       "1           NaN           NaN            NaN                NaN       NaN   \n",
       "2           NaN           NaN            NaN                NaN       NaN   \n",
       "3           NaN           NaN            NaN                NaN       NaN   \n",
       "4           NaN           NaN            NaN                NaN       NaN   \n",
       "...         ...           ...            ...                ...       ...   \n",
       "11760       NaN           NaN            NaN                NaN       NaN   \n",
       "11761       NaN           NaN            NaN                NaN       NaN   \n",
       "11762       NaN           NaN            NaN                NaN       NaN   \n",
       "11763       NaN           NaN            NaN                NaN       NaN   \n",
       "11764       NaN           NaN            NaN                NaN       NaN   \n",
       "\n",
       "      gene_symbol7 gene_biotype7 gene_description7  \n",
       "0              NaN           NaN               NaN  \n",
       "1              NaN           NaN               NaN  \n",
       "2              NaN           NaN               NaN  \n",
       "3              NaN           NaN               NaN  \n",
       "4              NaN           NaN               NaN  \n",
       "...            ...           ...               ...  \n",
       "11760          NaN           NaN               NaN  \n",
       "11761          NaN           NaN               NaN  \n",
       "11762          NaN           NaN               NaN  \n",
       "11763          NaN           NaN               NaN  \n",
       "11764          NaN           NaN               NaN  \n",
       "\n",
       "[11765 rows x 71 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['CHROM', 'POS', 'REF', 'ALT', 'QUAL', 'start', 'end', 'alleles',\n",
       "       'affected_start', 'affected_end', 'BRF', 'FR', 'HP', 'HapScore', 'MGOF',\n",
       "       'MMLQ', 'MQ', 'NF', 'NR', 'PP', 'QD', 'SC', 'SbPval', 'Source', 'TC',\n",
       "       'TCF', 'TCR', 'TR', 'WE', 'WS', 'sample', 'GT', 'GL', 'GOF', 'GQ', 'NV',\n",
       "       'ALT_type', 'gene_id', 'gene_symbol', 'gene_biotype',\n",
       "       'gene_description', 'most_severe_consequence', 'gene_id1',\n",
       "       'gene_symbol1', 'gene_biotype1', 'gene_description1', 'FILTER',\n",
       "       'gene_id2', 'gene_symbol2', 'gene_biotype2', 'gene_description2',\n",
       "       'gene_id3', 'gene_symbol3', 'gene_biotype3', 'gene_description3',\n",
       "       'gene_id4', 'gene_symbol4', 'gene_biotype4', 'gene_description4',\n",
       "       'gene_id5', 'gene_symbol5', 'gene_biotype5', 'gene_description5',\n",
       "       'gene_id6', 'gene_symbol6', 'gene_biotype6', 'gene_description6',\n",
       "       'gene_id7', 'gene_symbol7', 'gene_biotype7', 'gene_description7'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate VAF\n",
    "out_df[\"VAF\"] = out_df[\"NV\"].astype(int) / out_df[\"NR\"].astype(int)\n",
    "\n",
    "# calculate reference reads\n",
    "out_df[\"NWT\"] = out_df[\"NR\"].astype(int) - out_df[\"NV\"].astype(int)\n",
    "\n",
    "# calculate reference %\n",
    "out_df[\"RAF\"] = out_df[\"NWT\"].astype(int) / out_df[\"NR\"].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SNV    10838\n",
       "MNV      891\n",
       "Name: ALT_type, dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_df[\"ALT_type\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "intron_variant                         3942\n",
       "missense_variant                       2549\n",
       "synonymous_variant                     2243\n",
       "non_coding_transcript_exon_variant     1184\n",
       "splice_polypyrimidine_tract_variant     521\n",
       "3_prime_UTR_variant                     299\n",
       "splice_region_variant                   270\n",
       "intergenic_variant                      262\n",
       "5_prime_UTR_variant                     109\n",
       "regulatory_region_variant                96\n",
       "splice_donor_region_variant              85\n",
       "frameshift_variant                       47\n",
       "inframe_deletion                         25\n",
       "inframe_insertion                        23\n",
       "splice_donor_5th_base_variant            20\n",
       "splice_acceptor_variant                  20\n",
       "stop_gained                              17\n",
       "splice_donor_variant                     14\n",
       "stop_lost                                11\n",
       "TF_binding_site_variant                  10\n",
       "start_lost                                9\n",
       "coding_sequence_variant                   7\n",
       "stop_retained_variant                     1\n",
       "mature_miRNA_variant                      1\n",
       "Name: most_severe_consequence, dtype: int64"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_df[\"most_severe_consequence\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# 3. Loop, with functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record(CHROM=1, POS=1158631, REF=A, ALT=[G])\n",
      "\n",
      "Record(CHROM=1, POS=1246004, REF=A, ALT=[G])\n",
      "\n",
      "Record(CHROM=1, POS=1249187, REF=G, ALT=[A])\n",
      "\n",
      "Record(CHROM=1, POS=1261824, REF=G, ALT=[C])\n",
      "\n",
      "Record(CHROM=1, POS=1387667, REF=C, ALT=[G])\n",
      "\n",
      "Record(CHROM=1, POS=1585597, REF=A, ALT=[G])\n",
      "\n",
      "Record(CHROM=1, POS=1585642, REF=G, ALT=[T])\n",
      "\n",
      "Record(CHROM=1, POS=1586752, REF=T, ALT=[C])\n",
      "\n",
      "Record(CHROM=1, POS=1647686, REF=A, ALT=[C])\n",
      "\n",
      "Record(CHROM=1, POS=1647722, REF=GCTGTGACA, ALT=[TCTAGGATG])\n",
      "\n",
      "\n",
      "Done! Total records: 10\n",
      "CPU times: user 331 ms, sys: 32.8 ms, total: 364 ms\n",
      "Wall time: 20.8 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "out = []\n",
    "vcf_read = vcf.Reader(open(\"test_vcf_data.txt\", \"r\"))\n",
    "\n",
    "for i, record in enumerate(vcf_read):\n",
    "    if i < 40:\n",
    "        print(record, end = \"\\n\\n\")\n",
    "        out.append(process_record(record, i))\n",
    "    else:\n",
    "        break\n",
    "\n",
    "print(\"\\nDone! Total records: {}\".format(i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_df = pd.concat(out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate VAF\n",
    "out_df[\"VAF\"] = out_df[\"NV\"].astype(int) / out_df[\"NR\"].astype(int)\n",
    "\n",
    "# calculate reference reads\n",
    "out_df[\"NWT\"] = out_df[\"NR\"].astype(int) - out_df[\"NV\"].astype(int)\n",
    "\n",
    "# calculate reference %\n",
    "out_df[\"RAF\"] = out_df[\"NWT\"].astype(int) / out_df[\"NR\"].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0            G\n",
       "1            G\n",
       "2            A\n",
       "3            C\n",
       "4            G\n",
       "5            G\n",
       "6            T\n",
       "7            C\n",
       "8            C\n",
       "9    TCTAGGATG\n",
       "Name: ALT, dtype: object"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_df.ALT"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "scanpy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
